File: docker/workers/python/worker.py
import json
import time
import pika

def callback(ch, method, properties, body):
    try:
        message = json.loads(body)
        event_type = message.get('type', 'unknown')
        print(f" [x] Outbox event received: {event_type}")
    except json.JSONDecodeError:
        print(f" [x] Outbox message received (non-JSON): {body.decode(errors='replace')}")
    finally:
        ch.basic_ack(delivery_tag=method.delivery_tag)

def main():
    while True:
        try:
            connection = pika.BlockingConnection(pika.ConnectionParameters(host='rabbitmq'))
            channel = connection.channel()

            channel.queue_declare(queue='outbox', durable=True)

            channel.basic_consume(queue='outbox', on_message_callback=callback)

            print(' [*] Python worker waiting for messages on "outbox". To exit press CTRL+C')
            channel.start_consuming()
        except pika.exceptions.AMQPConnectionError:
            print("Connection to RabbitMQ failed. Retrying in 5 seconds...")
            time.sleep(5)


if __name__ == '__main__':
    main()

File: docker/workers/go/worker.go
package main

import (
	"encoding/json"
	"log"
	"time"

	"github.com/streadway/amqp"
)

type message struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

func main() {
	for {
		conn, err := amqp.Dial("amqp://guest:guest@rabbitmq:5672/")
		if err != nil {
			log.Printf("Failed to connect to RabbitMQ: %s. Retrying in 5 seconds...", err)
			time.Sleep(5 * time.Second)
			continue
		}
		defer conn.Close()

		ch, err := conn.Channel()
		if err != nil {
			log.Printf("Failed to open a channel: %s", err)
			continue
		}
		defer ch.Close()

		q, err := ch.QueueDeclare(
			"outbox", // name
			true,     // durable
			false,    // delete when unused
			false,    // exclusive
			false,    // no-wait
			nil,      // arguments
		)
		if err != nil {
			log.Printf("Failed to declare a queue: %s", err)
			continue
		}

		msgs, err := ch.Consume(
			q.Name, // queue
			"",     // consumer
			false,  // auto-ack
			false,  // exclusive
			false,  // no-local
			false,  // no-wait
			nil,    // args
		)
		if err != nil {
			log.Printf("Failed to register a consumer: %s", err)
			continue
		}

		log.Printf(" [*] Go worker waiting for messages on %q. To exit press CTRL+C", q.Name)

		for d := range msgs {
			var msg message
			if err := json.Unmarshal(d.Body, &msg); err != nil {
				log.Printf("Outbox message received (non-JSON): %s", d.Body)
				_ = d.Ack(false)
				continue
			}
			log.Printf("Outbox event received: %s", msg.Type)
			_ = d.Ack(false)
		}
	}
}
